### PR TITLE
chore(pruning): Support `IS NOT NULL` predicates in `PruningPredicate`

### PR DESCRIPTION
**Pull request by appletreeisyellow**
## Which issue does this PR close?

Help with #9171 and #7869

## Rationale for this change

`PruningPredicate` rewrites the original predicate into an expression that references min/max/null_count values of each column in the original predicate

Currently, `IS NOT NULL` is not supported, which means the rewritten predicate is ALWAYS TRUE, e.g.: 

Original Predicate | Rewritten Predicate
---------------- | -------------
`x IS NOT NULL` | `true`

See the code below, `phys_expr::IsNotNullExpr` is not listed/supported:

https://github.com/apache/arrow-datafusion/blob/292865e38cba1275a0f863230d816ee0e08c1318/datafusion/core/src/physical_optimizer/pruning.rs#L1139-L1140

## What changes are included in this PR?

Support `IS NOT NULL` predicate rewrite, e.g.:

Original Predicate | Rewritten Predicate
---------------- | -------------
`x IS NOT NULL` | `x_null_count = 0`

## Are these changes tested?

Yes

## Are there any user-facing changes?

No API change. User will experience performance improvements on queries that include `IS NOT NULL` predicates

### CODE REVIEW

File: datafusion/core/src/datasource/physical_plan/parquet/row_groups.rs [hunk 1]
```diff
@@ -293,6 +293,7 @@ impl PruningStatistics for BloomFilterStatistics {
 ///
 /// Note: This should be implemented for an array of [`RowGroupMetaData`] instead
 /// of per row-group
+#[derive(Debug)]
```
Comment (viirya) [inline datafusion/core/src/datasource/physical_plan/parquet/row_groups.rs:296]: Do we need this additionally?

File: datafusion/core/src/datasource/physical_plan/parquet/row_groups.rs [hunk 2]
```diff
@@ -620,13 +620,20 @@ mod tests {
...
+            ],
+        );
+        vec![rgm1, rgm2, rgm3]
     }
 
     #[test]
     fn row_group_pruning_predicate_null_expr() {
         use datafusion_expr::{col, lit};
-        // int > 1 and IsNull(bool) => c1_max > 1 and bool_null_count > 0
+        // c1 > 15 and IsNull(c2) => c1_max > 15 and c2_null_count > 0
```
Comment (appletreeisyellow) [inline datafusion/core/src/datasource/physical_plan/parquet/row_groups.rs:636]: Update comment to reflect the examples given in the test

File: datafusion/core/src/datasource/physical_plan/parquet/row_groups.rs (hunk 1 above)
Comment (appletreeisyellow) [reply datafusion/core/src/datasource/physical_plan/parquet/row_groups.rs]: Nice catch! I just removed it

File: datafusion/core/src/datasource/physical_plan/parquet/row_groups.rs (hunk 2 above)
Comment (viirya) [reply datafusion/core/src/datasource/physical_plan/parquet/row_groups.rs:636]: 👍 This reads more smoothly.

File: datafusion/core/src/datasource/physical_plan/parquet/row_groups.rs [hunk 3]
```diff
@@ -657,7 +664,7 @@ mod tests {
         use datafusion_expr::{col, lit};
         // test row group predicate with an unknown (Null) expr
         //
-        // int > 1 and bool = NULL => c1_max > 1 and null
+        // c1 > 15 and c2 = NULL => c1_max > 15 and null
```
Comment (viirya) [inline datafusion/core/src/datasource/physical_plan/parquet/row_groups.rs:667]: nit (but as take this chance to change it):

```suggestion
        // c1 > 15 and c2 = NULL => c1_max > 15 and NULL
```

File: datafusion/core/src/physical_optimizer/pruning.rs [hunk 4]
```diff
@@ -315,6 +315,7 @@ pub trait PruningStatistics {
 /// `x < 5` | `x_max < 5`
 /// `x = 5 AND y = 10` | `x_min <= 5 AND 5 <= x_max AND y_min <= 10 AND 10 <= y_max`
 /// `x IS NULL`  | `x_null_count > 0`
+/// `x IS NOT NULL`  | `x_null_count = 0
```
Comment (viirya) [inline datafusion/core/src/physical_optimizer/pruning.rs:318]:
```suggestion
/// `x IS NOT NULL`  | `x_null_count = 0`
```

**Review by viirya — APPROVED** [review]

**Comment by appletreeisyellow** [comment]
> Can you please one more test of this logic integrated into the overall row group pruning logic (to make sure it works in the context of a predicate). I think you can follow the model of these tests for `IS NULL`:

@alamb I added one more test into the row group pruning logic in https://github.com/apache/arrow-datafusion/pull/9208/commits/5040cf836554fb6eae1920260302b6263be7829e

**Comment by appletreeisyellow** [comment]
Thank you @alamb and @viirya for the review!

**Comment by alamb** [comment]
I believe this PR has a logic bug that we did not catch in code review: https://github.com/apache/arrow-datafusion/issues/9230

**Pull request merged**